src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1837 @@
+
+    // Find index of a potential match
+    __ align(OptoLoopAlignment);
+    __ bind(L_innerLoop);
+    __ tzcntl(r11, mask);
+
+    __ leaq(rTmp, Address(haystack, r11, Address::times_1, isU ? 4 : 2));
+
+    // Check for needle equality.  Handles UU and LL cases since byte comparison should be exact
+    __C2 arrays_equals(false, rTmp, firstNeedleCompare, compLen, rTmp3, rTmp2, XMM_TMP3, XMM_TMP4,
+                       false /* char */, knoreg);
+    __ testl(rTmp3, rTmp3);
+    __ jne(L_returnR11);
+
+    // Restore saved registers
+    __ movq(compLen, saveCompLen);
+    __ movq(firstNeedleCompare, saveNeedleAddress);
+
+    // Jump to inner loop if more matches to check, otherwise return not found
+    CLEAR_BIT(mask);
+    __ jne(L_innerLoop);
+    __ jmp(L_returnError);
+
+#undef saveCompLen
+#undef saveNeedleAddress
+  }
+
+  if (isReallyUL) {
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //                         Wide char code
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Pseudo-code:
+    //
+    // If needle length less than MAX_NEEDLE_LEN_TO_EXPAND, read the needle
+    // bytes from r14 and write them as words onto the stack.  Then go to the
+    // "regular" UU code.  This is equavilent to doing a UU comparison, since the
+    // haystack will be in UTF-16.
+    //
+    // If the needle can't be expanded, process the same way as the default
+    // cases above.
+    __ bind(L_wcharBegin);
+
+    // Restore argument encoding from UU back to UL for helpers
+    ae = StrIntrinsicNode::UL;
+    isUL = true;
+    isUU = false;
+
+    Label L_top, L_finished;
+
+    const Register haystack = rdi;
+    const Register hsLen = rsi;
+    const Register needle = rdx;
+    const Register nLen = rcx;
+
+    const Register offset = rax;
+    const Register index = rbx;
+    const Register wr_index = r13;
+
+    assert(MAX_NEEDLE_LEN_TO_EXPAND >= 32, "Small UL needles not supported");
+
+    // haystack length to bytes
+    __ shlq(hsLen, 1);
+
+    // Ensure haystack >= needle
+    __ leaq(index, Address(nLen, nLen, Address::times_1));
+    __ cmpq(index, hsLen);
+    __ jg(L_returnError);
+
+    // Can't expand large-ish needles
+    __ cmpq(nLen, MAX_NEEDLE_LEN_TO_EXPAND);
+    __ ja(L_wideNoExpand);
+
+    //
+    // Reads of existing needle are 16-byte chunks
+    // Writes to copied needle are 32-byte chunks
+    // Don't read past the end of the existing needle
+    //
+    // Start first read at [((ndlLen % 16) - 16) & 0xf]
+    // outndx += 32
+    // inndx += 16
+    // cmp nndx, ndlLen
+    // jae done
+    //
+    // Final index of start of needle @((16 - (ndlLen %16)) & 0xf) << 1
```

Comment:
What is the meaning of the `@`? Maybe `at`. I'd use the same consistently

Reply:
Changed to "at".